File: Walking Robot Simulation/Walking Robot Simulation.cpp
// Solution for Walking Robot Simulation in CPP

class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {

        /*
        ============================================================
        STEP 1: Store obstacles for O(1) lookup
        ============================================================

        We convert (x, y) → "x_y" string for hashing
        */
        unordered_set<string> obstacleSet;

        for (const auto& obs : obstacles) {
            string key = to_string(obs[0]) + "_" + to_string(obs[1]);
            obstacleSet.insert(key);
        }

        /*
        ============================================================
        STEP 2: Initialize robot state
        ============================================================
        */
        int x = 0, y = 0;                 // current position
        int maxDistanceSquared = 0;       // answer

        /*
            Direction vector (dx, dy)

            Initially facing NORTH → (0, +1)
        */
        pair<int, int> direction = {0, 1};

        /*
        ============================================================
        STEP 3: Process each command
        ============================================================
        */
        for (int command : commands) {

            /*
            ------------------------------------------------------------
            TURN LEFT (90°)
            (dx, dy) → (-dy, dx)
            ------------------------------------------------------------
            */
            if (command == -2) {
                direction = {-direction.second, direction.first};
            }

            /*
            ------------------------------------------------------------
            TURN RIGHT (90°)
            (dx, dy) → (dy, -dx)
            ------------------------------------------------------------
            */
            else if (command == -1) {
                direction = {direction.second, -direction.first};
            }

            /*
            ------------------------------------------------------------
            MOVE FORWARD step-by-step
            ------------------------------------------------------------
            */
            else {

                /*
                    IMPORTANT:
                    Move ONE STEP at a time (not jump directly)

                    Because obstacle may be in between
                */
                while (command--) {

                    int nextX = x + direction.first;
                    int nextY = y + direction.second;

                    string key = to_string(nextX) + "_" + to_string(nextY);

                    /*
                        If obstacle → STOP movement immediately
                    */
                    if (obstacleSet.count(key)) break;

                    /*
                        Otherwise move
                    */
                    x = nextX;
                    y = nextY;
                }
            }

            /*
            ------------------------------------------------------------
            Update max distance from origin
            ------------------------------------------------------------
            */
            maxDistanceSquared = max(maxDistanceSquared, x * x + y * y);
        }

        return maxDistanceSquared;
    }
};

File: Walking Robot Simulation/Walking Robot Simulation.py
# Solution for Walking Robot Simulation in PY

class Solution:
    def robotSim(self, commands, obstacles):

        # STEP 1: Store obstacles
        obstacle_set = set()
        for x, y in obstacles:
            obstacle_set.add((x, y))

        # STEP 2: Initialize
        x, y = 0, 0
        max_distance_squared = 0

        # Facing NORTH
        dx, dy = 0, 1

        # STEP 3: Process commands
        for command in commands:

            # Turn Left
            if command == -2:
                dx, dy = -dy, dx

            # Turn Right
            elif command == -1:
                dx, dy = dy, -dx

            # Move Forward
            else:
                for _ in range(command):
                    next_x = x + dx
                    next_y = y + dy

                    # Stop if obstacle
                    if (next_x, next_y) in obstacle_set:
                        break

                    x, y = next_x, next_y

            # Update max distance
            max_distance_squared = max(max_distance_squared, x*x + y*y)

        return max_distance_squared
